### ISSUE TITLE
failed to switch Protocols 400 when running behind GCP ALB

### ISSUE THREAD

**Issue by jamesagarside** [gravitational/teleport#31202]
Expected behavior:
When logging in via `tsh` I get the error `failed to switch Protocols 400`, I would expect this to login.

Current behavior:
```
tsh login --proxy teleport.example.co --auth github
If browser window does not open automatically, open it by clicking on the link:
 http://127.0.0.1:52164/xxxxxxxxxxxxxxxxxxxxxxxx
ERROR: Unable to connect to ssh proxy at teleport.example.co:443. Confirm connectivity and availability.
failed to switch Protocols 400
```
I can login successfully via browser. 

Bug details:
- Teleport version: 13.3.5
- Debug logs
### Helm Values
```
acme: false
service.type: NodePort
proxyListenerMode: multiplex
ingress:
  enabled: "true"
  proxyListenerMode: "multiplex"
  service.type: "LoadBalancer"
annotations:
  ingress:
    ingress.gcp.kubernetes.io/pre-shared-cert: "teleport-cert"
    kubernetes.io/ingress.allow-http: "false"
    kubernetes.io/ingress.class: "gce"
  service: 
    cloud.google.com/app-protocols: '{"tls":"HTTPS"}'
    cloud.google.com/backend-config: '{"default": "teleport-backend-config"}'
proxy:
  teleportConfig:
    proxy_service:
      trust_x_forwarded_for: "false"
```
The backend config just sets the health check path and protocol.

### WebAPI Ping
```
curl https://teleport.example.co/webapi/ping | jq . 
{
  "auth": {
    "type": "local",
    "second_factor": "on",
    "preferred_local_mfa": "webauthn",
    "allow_passwordless": true,
    "allow_headless": true,
    "local": {
      "name": ""
    },
    "webauthn": {
      "rp_id": "teleport.example.co"
    },
    "private_key_policy": "none",
    "device_trust_disabled": true,
    "device_trust": {
      "disabled": true
    },
    "has_motd": false,
    "default_session_ttl": "12h0m0s"
  },
  "proxy": {
    "kube": {
      "enabled": true,
      "listen_addr": "0.0.0.0:3080"
    },
    "ssh": {
      "listen_addr": "0.0.0.0:3080",
      "tunnel_listen_addr": "0.0.0.0:3080",
      "web_listen_addr": "0.0.0.0:3080",
      "public_addr": "teleport.example.co:443"
    },
    "db": {
      "postgres_listen_addr": "0.0.0.0:3080",
      "mysql_listen_addr": "0.0.0.0:3080"
    },
    "tls_routing_enabled": true,
    "assist_enabled": false
  },
  "server_version": "13.3.5",
  "min_client_version": "12.0.0",
  "cluster_name": "teleport.example.co",
  "automatic_upgrades": false
}
```

**Comment by jamesagarside** [comment]
I've upgraded to v13.3.7 and the problem still persists. 
Here is some debug output
```
tsh login --proxy teleport.example.co --auth github -v -d
2023-08-31T11:17:10+01:00 DEBU [TSH]       Web proxy port was not set. Attempting to detect port number to use. tsh/tsh.go:3887
2023-08-31T11:17:10+01:00 DEBU [TSH]       Resolving default proxy port (insecure: false) tsh/resolve_default_addr.go:111
2023-08-31T11:17:10+01:00 DEBU [TSH]       Trying teleport.example.co:3080... tsh/resolve_default_addr.go:99
2023-08-31T11:17:10+01:00 DEBU [TSH]       Trying teleport.example.co:443... tsh/resolve_default_addr.go:99
2023-08-31T11:17:10+01:00 DEBU [TSH]       Address teleport.example.co:443 succeeded. Selected as canonical proxy address tsh/resolve_default_addr.go:191
2023-08-31T11:17:10+01:00 DEBU [TSH]       Waiting for all in-flight proxy address tests to finish tsh/resolve_default_addr.go:139
2023-08-31T11:17:10+01:00 DEBU [TSH]       Proxy address test failed error:[Get "https://teleport.example.co:3080/webapi/ping": context canceled] tsh/resolve_default_addr.go:63
2023-08-31T11:17:10+01:00 INFO [CLIENT]    No teleport login given. defaulting to user client/api.go:1052
2023-08-31T11:17:10+01:00 INFO [CLIENT]    no host login given. defaulting to user client/api.go:1062
2023-08-31T11:17:10+01:00 INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/private/tmp/com.apple.launchd.Ozc8H7yzeo/Listeners" client/api.go:4310
2023-08-31T11:17:10+01:00 DEBU [TSH]       Pinging the proxy to fetch listening addresses for non-web ports. tsh/tsh.go:3362
2023-08-31T11:17:10+01:00 DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.co:443 client/api.go:4265
2023-08-31T11:17:10+01:00 DEBU             Attempting GET teleport.example.co:443/webapi/ping/github webclient/webclient.go:129
2023-08-31T11:17:10+01:00 DEBU             ALPN connection upgrade required for "teleport.example.co:443": true. client/alpn_conn_upgrade.go:95
2023-08-31T11:17:10+01:00 DEBU [CLIENT]    Attempting to login with a new RSA private key. client/api.go:3644
2023-08-31T11:17:11+01:00 DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.co:443 client/api.go:4265
2023-08-31T11:17:11+01:00 DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.co:443, insecure=false, extraHeaders=map[]) client/weblogin.go:331
2023-08-31T11:17:11+01:00 INFO [CLIENT]    Waiting for response at: http://127.0.0.1:57304. client/redirect.go:157
If browser window does not open automatically, open it by clicking on the link:
 http://127.0.0.1:57304/6c6ade99-1fd7-4aa7-9385-574616f1b9f8
2023-08-31T11:17:12+01:00 DEBU [CLIENT]    Got response from browser. client/weblogin.go:419
2023-08-31T11:17:12+01:00 DEBU [KEYSTORE]  Teleport TLS certificate valid until "2023-08-31 22:16:34 +0000 UTC". client/client_store.go:106
2023-08-31T11:17:12+01:00 DEBU [KEYAGENT]  Deleting obsolete stored key with index {ProxyHost:teleport.example.co Username:user ClusterName:teleport.example.co}. client/keyagent.go:527
2023-08-31T11:17:12+01:00 DEBU [KEYSTORE]  Adding known host teleport.example.co with proxy teleport.example.co client/trusted_certs_store.go:393
2023-08-31T11:17:12+01:00 INFO [KEYAGENT]  Loading SSH key for user "user" and cluster "teleport.example.co". client/keyagent.go:195
2023-08-31T11:17:12+01:00 DEBU [KEYSTORE]  Teleport TLS certificate valid until "2023-08-31 22:17:12 +0000 UTC". client/client_store.go:106
2023-08-31T11:17:12+01:00 DEBU [KEYSTORE]  Teleport TLS certificate valid until "2023-08-31 22:17:12 +0000 UTC". client/client_store.go:106
2023-08-31T11:17:12+01:00 INFO [CLIENT]    Connecting to proxy=teleport.example.co:443 login="user" using TLS Routing client/api.go:3050
2023-08-31T11:17:12+01:00 DEBU [KEYSTORE]  Teleport TLS certificate valid until "2023-08-31 22:17:12 +0000 UTC". client/client_store.go:106
2023-08-31T11:17:12+01:00 DEBU [HTTP:PROX] No proxy set in environment, returning direct dialer. proxy/proxy.go:195
2023-08-31T11:17:12+01:00 DEBU             ALPN connection upgrade for teleport.example.co:443. client/alpn_conn_upgrade.go:187

ERROR REPORT:
Original Error: trace.aggregate failed to switch Protocols 400
Stack Trace:
	github.com/gravitational/teleport/api@v0.0.0/client/alpn_conn_upgrade.go:201 github.com/gravitational/teleport/api/client.(*alpnConnUpgradeDialer).DialContext
	github.com/gravitational/teleport/api@v0.0.0/client/contextdialer.go:227 github.com/gravitational/teleport/api/client.NewDialer.func1
	github.com/gravitational/teleport/api@v0.0.0/client/contextdialer.go:185 github.com/gravitational/teleport/api/client.NewDialer.tracedDialer.func2
	github.com/gravitational/teleport/api@v0.0.0/client/contextdialer.go:127 github.com/gravitational/teleport/api/client.ContextDialerFunc.DialContext
	github.com/gravitational/teleport/api@v0.0.0/client/alpn.go:140 github.com/gravitational/teleport/api/client.(*ALPNDialer).DialContext
	github.com/gravitational/teleport/lib/utils/proxy/proxy.go:71 github.com/gravitational/teleport/lib/utils/proxy.directDial.DialTimeout
	github.com/gravitational/teleport/lib/utils/proxy/proxy.go:58 github.com/gravitational/teleport/lib/utils/proxy.directDial.Dial
	github.com/gravitational/teleport/lib/client/api.go:3132 github.com/gravitational/teleport/lib/client.makeProxySSHClientWithTLSWrapper
	github.com/gravitational/teleport/lib/client/api.go:3051 github.com/gravitational/teleport/lib/client.makeProxySSHClient
	github.com/gravitational/teleport/lib/client/api.go:2998 github.com/gravitational/teleport/lib/client.(*TeleportClient).connectToProxy
	github.com/gravitational/teleport/lib/client/api.go:2975 github.com/gravitational/teleport/lib/client.(*TeleportClient).ConnectToProxy.func1
	runtime/asm_arm64.s:1197 runtime.goexit
User Message: Unable to connect to ssh proxy at teleport.example.co:443. Confirm connectivity and availability.
	failed to switch Protocols 400
```

**Comment by kmmanto** [comment]
@jamesagarside GCP docs indicate that the ALB will not propagate (send?) the `Upgrade` headers to the backend?

https://cloud.google.com/load-balancing/docs/https#target-proxies
```
Note: In accordance with [RFC 2616](https://datatracker.ietf.org/doc/html/rfc2616#page-92), the following hop-by-hop headers are not propagated by the target proxy: Connection, Keep-Alive, Proxy-Authenticate, Proxy-Authorization, TE, Trailers, Transfer-Encoding, and Upgrade.
```

**Comment by jamesagarside** [comment]
Ahh good spot.
Would Teleport not fall back to the X-Teleport-Upgrade header in scenarios like this in accordance with this PR? #29683

**Comment by kmmanto** [comment]
@jamesagarside Not if the LB drops the connection altogether. We need confirmation from GCP folks.

**Comment by greedy52** [comment]
@kmmanto yes seems you found the curl test that checks whether the our current "upgrade" implementation works for a load balancer or reverse proxy, as documented [here](https://goteleport.com/docs/architecture/tls-routing/#will-tls-routing-work-behind-my-reverse-proxy)
 
We are looking into using native websockets:
https://github.com/gravitational/teleport/issues/30493

We will test out GCP ALB when working on that.

**Comment by zmb3** [comment]
@greedy52 can this be closed?

**Comment by greedy52** [comment]
GCP ALB support added in https://github.com/gravitational/teleport/pull/36343 with WebSocket support, released in [v15.1.0](https://github.com/gravitational/teleport/releases/tag/v15.1.0)

Note `trust_x_forwarded_for` still won't work for GCP ALB when multiple IPs are forwarded from the ALB, as current implementation limits only one IP is present in the x-forwarded-for header.
